Adapter calculate and retrieve total spendable amount

simzzz opened the pull request:
Implemented `adapter.get_deposit(...)`for the Ethereum and Dummy adapters, including an integration test.

## Change Log:

- `adapter::EthereumAdapter` - Added a `get_deposit()` function which calculates and returns total spendable amount, using the `OUTPACE` and `ERC20` ABI's.
- `adapter::ethereum::Error` - Added two new possible errors. The first being `TokenNotWhitelisted` which is returned when the function is called with a channel with a deposit asset not being in the config's whitelist. The second is `InvalidDepositAsset` which is used when a channel's deposit asset cannot be converted to `Address`. This error is here temporarily and I assume it will be gone soon when the `channel_v5` replaces the old `channel` in the adapter.
- `dev.toml and prod.toml configs` - Changed the structure of `token_addr_whitelist` from an array of strings, to an array with objects that contain `address`, `min_token_units_for_deposit` and `precision`.
- `primitives::adapter::Deposit` - new struct representing the `get_deposit()` output. Added signature for `get_deposit` in the `impl`
- `primitives::channel_validator::asset_listed()` - Changed `asset_listed()` function to work with the new token whitelist structure from String Vec to HashMap.
- `primitives::config` - Changed token whitelist from String Vec to HashMap. Added a new struct called `TokenInfo` which represents the value for the token whitelist HashMap.

simzzz committed:
early draft of get_spendable() function

simzzz committed:
changed a wrong parameter

simzzz committed:
change the way minimum deposit is retrieved

simzzz committed:
progress made for ethereum adapter

simzzz committed:
updated /lib/protocol-eth

simzzz committed:
Done with Ethereum Adapter

simzzz committed:
Better error for TokenNotWhitelisted

elpiel commented on primitives/src/adapter.rs:
```diff
@@ -6,11 +6,17 @@ use serde::{Deserialize, Serialize};
 
 pub trait AdapterErrorKind: fmt::Debug + fmt::Display {}
 
+pub struct SpendableOutput {
```
I think that just `Spendable` is clear enough.

I am not 100% sure if `U256` will remain as type here, we will either use a `BigNum` and include the token info (like the precision) or we will use a `UnifiedNum` (precision 8) #384

elpiel commented on primitives/src/channel_validator.rs:
```diff
@@ -96,11 +97,11 @@ pub fn creator_listed(channel: &Channel, whitelist: &[ValidatorId]) -> bool {
 }
 
-pub fn asset_listed(channel: &Channel, whitelist: &[String]) -> bool {
+pub fn asset_listed(channel: &Channel, whitelist: &HashMap<String, TokenInfo>) -> bool {
```
When #382 is merged we can use the `Address` type instead of `String`, or we can use `ValidatorId` for the time being and later change it once the PR is merged.

elpiel commented on primitives/src/config.rs:
```diff
@@ -14,6 +15,12 @@ lazy_static! {
 
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct TokenInfo {
+    pub min_deposit: u32,
```
I am not sure if this will be enough. It matches ideally for DAI (18 decimals) since a u32::MAX is `4_294_967_295` (i.e. minimum deposit that will convert to `42` in 8 decimal precision). Maybe we can use BigNum instead (this means that the config value should be a string)?

elpiel commented on adapter/src/ethereum/error.rs:
```diff
@@ -22,6 +22,7 @@ pub enum Error {
     ContractQuerying(web3::contract::Error),
     /// Error occurred during verification of Signature and/or StateRoot and/or Address
     VerifyAddress(VerifyError),
+    TokenNotWhitelisted(String),
```
This should probably also change to `Address` when we merge the other PR

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +271,64 @@ impl Adapter for EthereumAdapter {
+
+        let erc20_contract = Contract::from_json(
+            self.web3.eth(),
+            self.config.ethereum_core_address.into(),
```
this should be the Token address instead of the core one.

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +271,64 @@ impl Adapter for EthereumAdapter {
+        let mut amount: U256 = tokio_compat_02::FutureExt::compat(async {
+            tokio_compat_02::FutureExt::compat(outpace_contract.query(
+                "deposits",
+                (channel.id.into_token(), spender.to_hex_prefix_string().into_token()),
```
You need these types to be from the following list:

https://docs.rs/web3/0.15.0/web3/contract/tokens/trait.Tokenize.html?search=into_token

- The `ChannelId` has `[u8; 32]` (`FixedBytes` or known as `bytes32` in the contact )
- The `ValidatorId` hsa `[u8; 20]` (or `Address`)
You can also see those types in the contract:

https://github.com/AdExNetwork/adex-protocol-eth/blob/identity-tests/contracts/OUTPACE.sol#L43
It's possible that the this will fail if you try to test it.

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +271,64 @@ impl Adapter for EthereumAdapter {
+        let to_be_deposited: U256 = tokio_compat_02::FutureExt::compat(async {
+            tokio_compat_02::FutureExt::compat(erc20_contract.query(
+                "balanceOf",
+                channel.deposit_asset.clone().into_token(),
```
This should probably also change, because we use `Address` for the token address in the new `Channel`.
The types right now are different, the contract expects an `Address`, however we pass it a `Token::String`

You can also check a more human-readable form of the ABI here: https://ethereumdev.io/abi-for-erc20-contract-on-ethereum/

elpiel reviewed (commented):
Looks like you have great progress on the task already :tada: 
Can you also make sure to include a list of the Changelog in PR so we can later add it to the CHANGELOG.md file.

elpiel commented:
After some discussion with Ivo, we came to the following changes:
`get_deposit(channel: &Channel, user: &Address) -> Deposit`
`struct Deposit { total: BigNum, pending: BigNum }`

This could change in the future, as we will introduce the `UnifiedNum` and we have yet to see if we will use the full precision from the validator (`BigNum`) or directly work with the `UnifiedNum` from the `Adapter` onward.

simzzz committed:
added Deposit type + fixes

simzzz committed:
Merge branch 'aip-61-adex-v5' into issue-380-adapter-total-spendable-amount

simzzz committed:
Merged new Channel/Campaign PR + more fixes

simzzz committed:
changed total + pending variables to be BigNum

simzzz committed:
Done with deserialization + tests passing + cargo fmt/clippy

simzzz committed:
renaming + updating protocol-eth

simzzz committed:
updated submodules

simzzz committed:
minor changes

simzzz committed:
fixed some unwraps

simzzz committed:
Simplified deserialization of token whitelist

simzzz committed:
Merge branch 'aip-61-adex-v5' into issue-380-adapter-total-spendable-amount

simzzz committed:
binded import to a different name + merged main v5 branch

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +273,77 @@ impl Adapter for EthereumAdapter {
+
+    async fn get_deposit(
+        &self,
+        channel: &Channel,
```
you can now use the new channel_v5::Channel`

elpiel commented on primitives/Cargo.toml:
```diff
@@ -54,6 +54,7 @@ futures = "0.3"
 async-trait = "0.1"
 # Other
 lazy_static = "1.4.0"
+web3 = { version = "0.15" }
```
do we use web3 in primitives? If now we can remove it.

elpiel commented on primitives/src/config.rs:
```diff
@@ -34,14 +44,44 @@ pub struct Config {
+    let mut tokens_whitelist = HashMap::new();
+    let array: Vec<ConfigWhitelist> = Deserialize::deserialize(deserializer)?;
+
+    array.into_iter().for_each(|i| {
```
you can actually `collect()` a tuple of `(config_whitelist.address, token_info)` directly into a `HashMap`. Small trick to remove the need of insertion, new mut hashmap and etc.
You would also need to use `.map()` instead of `for_each`.
Another option is `fold()`, but in this case `.map().collect()` will be least amount code.

elpiel commented on primitives/src/config.rs:
```diff
@@ -34,14 +44,44 @@ pub struct Config {
+struct ConfigWhitelist {
+    address: Address,
+    min_token_units_for_deposit: BigNum,
+    precision: NonZeroU8,
```
could you make `precision` & `decimals` have the same name? this will help out with confusion.
Also `min token units for deposit` might be a good idea to be in `UnifiedNum` since the only place where we would use BigNum should be in the merkle tree & adapter

elpiel reviewed (commented):
Can you also make sure to fix the gitmodule

simzzz committed:
adde ERC20.json to the adapter temporarily

simzzz committed:
formatting and removed unwrap

simzzz committed:
Fixes from PR review

simzzz committed:
Removed whitelist for dev.toml causing a test to fail

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +278,71 @@ impl Adapter for EthereumAdapter {
+            tokio_compat_02::FutureExt::compat(outpace_contract.query(
+                "deposits",
+                (
+                    H256(*channel.id()).into_token(),
```
this is not really readable, could you used the `Token::Address`, for the ChannelId we could leave it as it is. It wont really give us any benefit to use `Token::FixedBytes` (which is the equivalent of bytes32, bytes64, etc. in solidity)

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +278,71 @@ impl Adapter for EthereumAdapter {
+
+        let mut total = BigNum::from_str(&total.to_string())?;
+
+        let still_on_create_2: H256 = tokio_compat_02::FutureExt::compat(async {
```
Should be https://docs.rs/ethabi/13.0.0/ethabi/type.Uint.html

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +278,71 @@ impl Adapter for EthereumAdapter {
+        let still_on_create_2: H256 = tokio_compat_02::FutureExt::compat(async {
+            tokio_compat_02::FutureExt::compat(erc20_contract.query(
+                "balanceOf",
+                Bytes(channel.token.as_bytes().to_vec()).into_token(),
```
`EthAddress` (you can use the `channel.token.to_bytes()` which gives you `[u8; 20]`)

elpiel commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +278,71 @@ impl Adapter for EthereumAdapter {
+        .await
+        .map_err(Error::ContractQuerying)?;
+
+        let still_on_create_2 = BigNum::from_str(&still_on_create_2.to_string())?;
```
you can use `still_on_create_2.to_string().parse()` this is what `FromStr` allows you to do. You can use it directly like this, but it's meant more like a syntax sugar for String & &str `.parse()`

elpiel commented on adapter/src/ethereum/error.rs:
```diff
@@ -1,5 +1,6 @@
 use primitives::adapter::{AdapterErrorKind, Error as AdapterError};
-use primitives::ChannelId;
+use primitives::address::Error as AddressError;
```
can you group them together from now on?
E.g. `primitives::{ all imports here }`;

elpiel commented on docs/config/dev.toml:
```diff
@@ -26,5 +26,5 @@ ethereum_adapter_relayer = 'https://goerli-relayer.adex.network'
 minimal_fee = "0"
-token_address_whitelist = []
 validators_whitelist = []
+token_address_whitelist = []
```
please add tokens to the whitelist for the known used stablecoins we use: DAI, USDT, USDC.
Also if this is a Table could you please use the alternative styling of:
```
[[token_address_whitelist]]
address = ...
min_token_units_for_deposit = ...
...
[[token_address_whitelist]]
address = ...
min_token_units_for_deposit = ...
...

```

elpiel commented on docs/config/prod.toml:
```diff
@@ -27,3 +27,4 @@ minimal_deposit = "0"
 minimal_fee = "0"
 token_address_whitelist = ['0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359', '0x6B175474E89094C44Da98b954EedeAC495271d0F']
```
shouldn't this be dumped now?

elpiel commented on adapter/src/dummy.rs:
```diff
@@ -129,4 +130,16 @@ impl Adapter for DummyAdapter {
+        channel: &ChannelV5,
+        address: &Address,
+    ) -> AdapterResult<Deposit, Self::AdapterError> {
+        // TODO - complete function
```
any ideas on how we should do this in the DummyAdapter?

elpiel reviewed (changes requested):
Also I don't see any tests for the new method of the ethereum adapter

simzzz committed:
more fixes from PR

simzzz committed:
dummy adapter + integration tests

simzzz committed:
cargo fmt + realistic test spender address

simzzz committed:
comitting progress with smart contracts

elpiel committed:
Merge branch 'aip-61-adex-v5' into merge-aip-61-into-issue-380

elpiel committed:
fix merge issues

elpiel committed:
Merge pull request #397 from AdExNetwork/merge-aip-61-into-issue-380

Merge aip 61 into issue 380

simzzz committed:
changes to the dummy adapter

simzzz committed:
Merge branch 'aip-61-adex-v5' of https://github.com/AdExNetwork/adex-validator-stack-rust into aip-61-adex-v5

simzzz committed:
Update protocol-eth

simzzz committed:
Merge branch 'aip-61-adex-v5' into issue-380-adapter-total-spendable-amount

simzzz committed:
changes from merge

simzzz committed:
Merge branch 'issue-380-adapter-total-spendable-amount' of https://github.com/AdExNetwork/adex-validator-stack-rust into issue-380-adapter-total-spendable-amount

simzzz committed:
fixed salt

samparsky commented on adapter/src/ethereum.rs:
```diff
@@ -269,6 +296,69 @@ impl Adapter for EthereumAdapter {
+
+        let mut total = BigNum::from_str(&total.to_string())?;
+
+        let counterfactual_address = get_counterfactual_address(channel.token);
```
The counterfactual address doesn't depend on the `channel.token`, this should be `address` param

samparsky commented on adapter/src/ethereum.rs:
```diff
@@ -82,6 +91,24 @@ impl EthereumAdapter {
+    result.update(token.as_bytes());
+    result.update(&salt);
+    result.update(&hashed_init_code);
+    let mut res: [u8; 20] = [0; 20];
```
we need from the `12th byte downward` and not the `first 20 bytes`. You can checkout the specification here https://eips.ethereum.org/EIPS/eip-1014

samparsky commented on adapter/src/ethereum.rs:
```diff
@@ -82,6 +91,24 @@ impl EthereumAdapter {
     }
 }
 
+fn get_counterfactual_address(token: Address) -> H160 {
```
This requires some changes

This should accept the`channel struct`, `outpace address` and `depositor`.
We then `abi.encode` the params and append it to the bytecode of the depositor contract here https://github.com/AdExNetwork/adex-protocol-eth/blob/master/contracts/deposits/Depositor.sol that's what forms the `initcode` 

You can check out this article for more clarification https://medium.com/@martinetlee/notes-on-eip-1014-create2-1-351c47e22a89

samparsky replied:
You can also checkout this library https://docs.rs/create2/0.0.2/create2/fn.calc_addr_with_hash.html the address param there is the `sweeper` address

samparsky replied:
To properly test we would have to deploy the sweeper contract too

simzzz committed:
create2 address should be generated successfully right now

simzzz committed:
removed obsolete comment

simzzz committed:
Added additional test

simzzz committed:
uncommented code that was commented out while testing

simzzz committed:
changes to bytecode files and tests

elpiel committed:
scripts - ethereum - ganache-cli - setup accs with priv keys

elpiel committed:
Merge branch 'aip-61-adex-v5' into issue-380-adapter-total-spendable-amount

elpiel committed:
primitives - Config - add outpace & sweeper addresses
- primitives - Config - deprecate ethereum_core_address

elpiel committed:
adapter - Ethereum - fix get_deposit & test it
- primitives - ChannelId - fn as_bytes
- fixed Cargo.lock

elpiel committed:
- Address - fix var name

elpiel committed:
git - update protocol-eth submodule

elpiel committed:
adapter - use `protocol-eth` ABI & bytecode paths
- adapter - ABI & bytecode files clean up

elpiel commented:
## TODO:
- `get_deposit` of Dummy adapter

elpiel committed:
adapter - ethereum - fix outpace & sweeper bytecode paths for tests

elpiel committed:
rustfmt

samparsky commented on docs/config/dev.toml:
```diff
@@ -20,11 +20,30 @@ ip_rate_limit = { type = 'ip', timeframe = 20000 }
 ethereum_core_address = '0x333420fc6a897356e69b62417cd17ff012177d2b'
+# TODO: Replace with real contract address
+outpace_address = '0x333420fc6a897356e69b62417cd17ff012177d2b'
+# TODO: Replace with real contract address
```
to make tests reproducible we can create a ganache snapshot of the required deployments and use it on startup

elpiel replied:
This is a cool idea for enhancement!
How can we do that?

elpiel committed:
adapter - ethereum - add Sweeper contract sweep test case

elpiel committed:
adapter - dummy - impl mocking of deposit calls

elpiel committed:
adapter - ethereum - move setup fns to `mod test_utils`

elpiel committed:
adapter - remove `EthereumChannel` struct

elpiel committed:
primitives - Adapter - remove `validate_channel`

elpiel committed:
adapter - cleanup old code

elpiel committed:
adapter - ethereum - contrafactual address clean up

elpiel committed:
adapter - remove ChannelValidator & clean up code
TODO: uncomment sweeper test case once we figure out what's wrong with the contract call

elpiel committed:
WIP try to fix sweeper_sweep

samparsky committed:
fix: sweeper test case

samparsky committed:
fix: remove println

samparsky committed:
fix: out of gaserrors

samparsky committed:
update: Outpace bytecode file

elpiel committed:
git - Update protocol-eth

elpiel committed:
Merge pull request #402 from samparsky/fix-sweeper-testcase

sweeper test case

elpiel committed:
update protocol-eth

elpiel committed:
remove empty file that Sam introduced

elpiel committed:
adapter - ethereum - remove todos
